File: mobile_new/lib/features/bookings/domain/entities/booking.dart
import 'package:equatable/equatable.dart';

/// Booking status enum
enum BookingStatus { pending, confirmed, completed, cancelled }

/// BookingItem in a booking
class BookingItem extends Equatable {
  final String id;
  final String serviceId;
  final String serviceName;
  final double price;
  final int quantity;
  final DateTime? selectedDate;
  final String? selectedSlot;

  const BookingItem({
    required this.id,
    required this.serviceId,
    required this.serviceName,
    required this.price,
    required this.quantity,
    this.selectedDate,
    this.selectedSlot,
  });

  @override
  List<Object?> get props => [
    id,
    serviceId,
    serviceName,
    price,
    quantity,
    selectedDate,
    selectedSlot,
  ];
}

/// Booking entity
class Booking extends Equatable {
  final String id;
  final String userId;
  final List<BookingItem> items;
  final double totalAmount;
  final BookingStatus status;
  final DateTime createdAt;
  final DateTime? completedAt;

  const Booking({
    required this.id,
    required this.userId,
    required this.items,
    required this.totalAmount,
    required this.status,
    required this.createdAt,
    this.completedAt,
  });

  @override
  List<Object?> get props => [
    id,
    userId,
    items,
    totalAmount,
    status,
    createdAt,
    completedAt,
  ];
}

/// Paginated booking response
class BookingListResponse extends Equatable {
  final List<Booking> bookings;
  final int total;
  final int page;
  final int limit;
  final int totalPages;

  const BookingListResponse({
    required this.bookings,
    required this.total,
    required this.page,
    required this.limit,
    required this.totalPages,
  });

  @override
  List<Object?> get props => [bookings, total, page, limit, totalPages];
}

File: backend/src/models/Booking.ts
import mongoose, { Document, Schema } from 'mongoose';

export type BookingStatus =
  | 'pending'
  | 'confirmed'
  | 'completed'
  | 'cancelled'
  | 'failed';

export type PaymentStatus = 'pending' | 'completed' | 'failed' | 'cancelled';

export interface IBookingItem {
  _id: mongoose.Types.ObjectId;
  serviceId: mongoose.Types.ObjectId;
  serviceTitle: string;   // Snapshot at booking time
  serviceImage: string;
  slotDate: string;
  slotTime: string;
  quantity: number;
  pricePerUnit: number;
  subtotal: number;
}

export interface IBooking extends Document {
  _id: mongoose.Types.ObjectId;
  bookingRef: string;
  userId: mongoose.Types.ObjectId;
  items: IBookingItem[];
  totalAmount: number;
  status: BookingStatus;
  paymentStatus: PaymentStatus;
  paymentId?: string;
  cancelledAt?: Date;
  cancellationReason?: string;
  notes?: string;
  createdAt: Date;
  updatedAt: Date;
}

const bookingItemSchema = new Schema<IBookingItem>(
  {
    serviceId: {
      type: Schema.Types.ObjectId,
      ref: 'Service',
      required: true,
    },
    serviceTitle: { type: String, required: true },
    serviceImage: { type: String, default: '' },
    slotDate: { type: String, required: true },
    slotTime: { type: String, required: true },
    quantity: { type: Number, required: true, min: 1 },
    pricePerUnit: { type: Number, required: true },
    subtotal: { type: Number, required: true },
  },
  { _id: true },
);

const bookingSchema = new Schema<IBooking>(
  {
    bookingRef: {
      type: String,
      required: true,
      unique: true,
    },
    userId: {
      type: Schema.Types.ObjectId,
      ref: 'User',
      required: true,
    },
    items: {
      type: [bookingItemSchema],
      required: true,
      validate: [(arr: IBookingItem[]) => arr.length > 0, 'Booking must have at least one item'],
    },
    totalAmount: {
      type: Number,
      required: true,
      min: 0,
    },
    status: {
      type: String,
      enum: ['pending', 'confirmed', 'completed', 'cancelled', 'failed'],
      default: 'pending',
    },
    paymentStatus: {
      type: String,
      enum: ['pending', 'completed', 'failed', 'cancelled'],
      default: 'pending',
    },
    paymentId: { type: String },
    cancelledAt: { type: Date },
    cancellationReason: { type: String, maxlength: 500 },
    notes: { type: String, maxlength: 500 },
  },
  {
    timestamps: true,
    toJSON: { virtuals: true, versionKey: false },
  },
);

bookingSchema.index({ userId: 1, createdAt: -1 });
bookingSchema.index({ bookingRef: 1 });
bookingSchema.index({ status: 1 });

export const Booking = mongoose.model<IBooking>('Booking', bookingSchema);
